samples/snippets/sample_templates.py

```diff
@@ -0,0 +1,195 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Iterable
+
+from google.cloud import compute_v1
+
+
+def get_instance_template(project_id: str, template_name: str) -> compute_v1.InstanceTemplate:
+    """
+    Retrieve information about given instance template.
+
+    Args:
+        project_id: project ID or project number of the Cloud project you use.
+        template_name: name of the template you want to get.
+
+    Returns:
+        InstanceTemplate object containing information about given template.
+    """
+    template_client = compute_v1.InstanceTemplatesClient()
+    return template_client.get(project=project_id, instance_template=template_name)
+
+
+def list_instance_templates(project_id: str) -> Iterable[compute_v1.InstanceTemplate]:
+    """
+    Get a list of InstanceTemplate objects available in given project.
+
+    Args:
+        project_id: project ID or project number of the Cloud project you use.
+
+    Returns:
+        Iterable list of InstanceTemplate objects.
+    """
+    template_client = compute_v1.InstanceTemplatesClient()
+    return template_client.list(project=project_id)
+
+
+def create_template(project_id: str, template_name: str) -> compute_v1.InstanceTemplate:
+    """
+    Create a new template in given project with provided name.
+
+    Args:
+        project_id: project ID or project number of the Cloud project you use.
+        template_name: name of the new template that is created.
+
+    Returns:
+        InstanceTemplate object describing the new instance template.
+    """
+    # Describe the size and source image of the boot disk to attach to the instance.
+    disk = compute_v1.AttachedDisk()
+    initialize_params = compute_v1.AttachedDiskInitializeParams()
+    initialize_params.source_image = (
+        "projects/debian-cloud/global/images/family/debian-11"
+    )
+    initialize_params.disk_size_gb = 250
+    disk.initialize_params = initialize_params
+    disk.auto_delete = True
+    disk.boot = True
+
+    # Use the default network, without specifying a subnetwork.
+    network_interface = compute_v1.NetworkInterface()
+    network_interface.name = "global/networks/default"
+
+    # Let the instance have external IP
```

Comment:
For external IP:
"The template lets the instance use an external IP address through Cloud NAT."

I attempted to clarify, but correct me if I'm misunderstanding the behavior here.

Reply:
That's not that simple and I'm not sure how to describe it well... ONE_TO_ONE_NAT actually means the Cloud NAT is not involved (that's a weird, I know). But regular NAT routing takes place to deliver the traffic.

See [here](https://cloud.google.com/nat/docs/overview):

>The Compute Engine VM's network interface's primary internal IP address, provided that the network interface doesn't have an external IP address assigned to it. If the network interface has an external IP address assigned to it, Google Cloud automatically performs one-to-one NAT for packets whose sources match the interface's primary internal IP address because the network interface meets the Google Cloud internet access requirements. The existence of an external IP address on an interface always takes precedence and always performs one-to-one NAT, without using Cloud NAT.

Cloud NAT is used when you have a bunch of VMs with *only internal* IPs and you want to give them Internet access. They can "share" one external IP then, to talk with the Internet and our Cloud NAT is making sure that the traffic is routed properly to those instances.

I'm honestly not sure about the origin on the ONE_TO_ONE_NAT value, why is it called like that... Perhaps it's because there's basically no address translation? The IP stays as is, so it's one-to-one like that?